### PR TITLE
Support start activities that contain an inner class

### PR DESCRIPTION
**Pull request by vikki**
Hopefully fixing appium/appium#8359 (more details there), whereby you can't specify an inner class as the start activity, because every thing from the dollar onwards gets clipped. Fix is just to escape the dollar (awfully similar to my last PR!)

### CODE REVIEW

File: lib/tools/apk-utils.js [hunk 1]
```diff
@@ -41,6 +41,8 @@ apkUtilsMethods.startApp = async function (startAppOptions = {}) {
       log.errorAndThrow("activity and pkg is required for launching application");
     }
     startAppOptions = _.clone(startAppOptions);
+    startAppOptions.activity = startAppOptions.activity.replace('$', '\\$');
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: shouldn't we check that activity property is defined first?

File: test/functional/apk-utils-e2e-specs.js [hunk 2]
```diff
@@ -131,6 +131,18 @@ describe('apk utils', function () {
...
+
+  it('should start activity when start activity is an inner class', async () => {
+    await adb.install(contactManagerPath);
+    await adb.startApp({pkg: 'com.android.settings',
+      activity: '.Settings$NotificationAppListActivity'});
+
+    let {appPackage, appActivity} = await adb.getFocusedPackageAndActivity();
+    appPackage.should.equal('com.android.settings');
+    appActivity.should.equal('.Settings$NotificationAppListActivity');
+
```
Comment (mykola-mokhnach) [inline test/functional/apk-utils-e2e-specs.js]: this empty line is not needed

File: test/unit/apk-utils-specs.js [hunk 3]
```diff
@@ -354,6 +354,19 @@ describe('Apk-utils', () => {
...
+      const startAppOptionsWithInnerClass = { pkg: 'pkg', activity: 'act$InnerAct'},
+            cmdWithInnerClass = ['am', 'start', '-W', '-n', 'pkg/act\\$InnerAct', '-S'];
+
+      mocks.adb.expects('getApiLevel')
+        .once().withExactArgs()
+        .returns('17');
+      mocks.adb.expects('shell')
+        .once().withExactArgs(cmdWithInnerClass)
+        .returns('');
+      (await adb.startApp(startAppOptionsWithInnerClass));
```
Comment (mykola-mokhnach) [inline test/unit/apk-utils-specs.js]: extra brackets here

File: test/unit/apk-utils-specs.js [hunk 4]
```diff
@@ -354,6 +354,19 @@ describe('Apk-utils', () => {
       (await adb.startApp(startAppOptions));
       mocks.adb.verify();
     });
+    it('should call getApiLevel and shell with correct arguments when activity is inner class', async () => {
+      const startAppOptionsWithInnerClass = { pkg: 'pkg', activity: 'act$InnerAct'},
+            cmdWithInnerClass = ['am', 'start', '-W', '-n', 'pkg/act\\$InnerAct', '-S'];
```
Comment (mykola-mokhnach) [inline test/unit/apk-utils-specs.js]: I'd rather define this var in a separate line for better readability

Comment (vikki) [inline test/unit/apk-utils-specs.js]: I think I did it that way to start with but the rest of the file uses this style, e.g. line 10. Still want me to change it (I don't mind, just trying to be consistent)?

File: test/unit/apk-utils-specs.js (hunk 3 above)
Comment (vikki) [inline test/unit/apk-utils-specs.js]: Ahh yes, they are unnecessary, but again consistent with the other tests in the block, e.g. l342 - should I remove?

File: lib/tools/apk-utils.js (hunk 1 above)
Comment (vikki) [inline lib/tools/apk-utils.js]: It's covered on l40, right? So it'll throw and never get this far if we don't have an activity defined.

File: test/unit/apk-utils-specs.js (hunk 3 above)
Comment (mykola-mokhnach) [inline test/unit/apk-utils-specs.js]: yup, remove pls

File: test/unit/apk-utils-specs.js (hunk 4 above)
Comment (mykola-mokhnach) [inline test/unit/apk-utils-specs.js]: Up to you

File: lib/tools/apk-utils.js (hunk 1 above)
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: yes, it's ok

**Comment by imurchie** [comment]
Thanks for this!

**Comment by imurchie** [comment]
Published in `appium-android-driver@1.19.0`.

**Pull request merged**